http.go

```diff
@@ -0,0 +1,107 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tspclient
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+)
+
+// maxBodyLength specifies the max content can be received from the possibly
+// malicious remote server.
+// The legnth of a regular TSA response with certificates is usually less than
+// 10 KiB.
+var maxBodyLength = 1 * 1024 * 1024 // 1 MiB
+
+// TimestampQuery is the content-type of timestamp query.
+// RFC 3161 3.4
+const TimestampQuery = "application/timestamp-query"
+
+// TimestampReply is the content-type of timestamp reply
+// RFC 3161 3.4
+const TimestampReply = "application/timestamp-reply"
+
+// httpTimestamper is a HTTP-based timestamper.
+type httpTimestamper struct {
+	rt       http.RoundTripper
+	endpoint string
+}
+
+// NewHTTPTimestamper creates a HTTP-based timestamper with the endpoint
+// provided by the TSA.
+// http.DefaultTransport is used if nil RoundTripper is passed.
+func NewHTTPTimestamper(rt http.RoundTripper, endpoint string) (Timestamper, error) {
+	if rt == nil {
+		rt = http.DefaultTransport
+	}
+	if _, err := url.Parse(endpoint); err != nil {
+		return nil, err
+	}
+	return &httpTimestamper{
+		rt:       rt,
+		endpoint: endpoint,
+	}, nil
+}
+
+// Timestamp sends the request to the remote TSA server for timestamping.
+// Reference: RFC 3161 3.4 Time-Stamp Protocol via HTTP
+func (ts *httpTimestamper) Timestamp(ctx context.Context, req *Request) (*Response, error) {
+	// prepare for http request
+	reqBytes, err := req.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	hReq, err := http.NewRequestWithContext(ctx, http.MethodPost, ts.endpoint, bytes.NewReader(reqBytes))
+	if err != nil {
+		return nil, err
+	}
+	hReq.Header.Set("Content-Type", TimestampQuery)
+
+	// send the request to the remote TSA server
+	hResp, err := ts.rt.RoundTrip(hReq)
+	if err != nil {
+		return nil, err
+	}
+	defer hResp.Body.Close()
+
+	// verify HTTP response
+	if hResp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%s %q: https response bad status: %s", http.MethodPost, ts.endpoint, hResp.Status)
+	}
+	if contentType := hResp.Header.Get("Content-Type"); contentType != TimestampReply {
+		return nil, fmt.Errorf("%s %q: unexpected response content type: %s", http.MethodPost, ts.endpoint, contentType)
+	}
+
+	// read response
+	lr := &io.LimitedReader{
+		R: hResp.Body,
+		N: int64(maxBodyLength),
+	}
+	respBytes, err := io.ReadAll(lr)
+	if err != nil {
+		return nil, err
+	}
+	if lr.N == 0 {
+		return nil, fmt.Errorf("%s %q: https response reached the %d MiB size limit", hResp.Request.Method, hResp.Request.URL, maxBodyLength/1024/1024)
+	}
+	var resp Response
+	if err := resp.UnmarshalBinary(respBytes); err != nil {
```

Comment:
Should we validate response before returning it? 
Validation should include
1. Nonce equality check
2. certReq value check to validate cert actually got added in response and vice versa
3. If any check(didnt read spec) related to extensions.

Reply:
@priteshbandi This is by design to decouple TSA timestamp requestor and timestamp verifier. The validation of the response happens on the timestamp token verification side, right before the verifier consume the timestamp. You will see this logic in `notation-core-go` PR once it's ready. The check also includes Notation specific checks such as TSA signing certificate chain validation, algorithm validation, etc.

Reply:
>The validation of the response happens on the timestamp token verification side, right before the verifier consume the timestamp. 
How will you verify the response against request?

Reply:
Ah, I misunderstood the question. The timestamp token verification is done in notation-go side. Adding the check of response in.

Reply:
Code updated.